utils/src/qrcode.rs

```diff
@@ -0,0 +1,248 @@
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub const FILLED_CHAR: char = '█';
+pub const EMPTY_CHAR: char = ' ';
+pub const NEW_LINE: char = '\n';
+
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum QrCodeError {
+    #[error("Given data is too long to fit in a QR code: {0}")]
+    DataTooLong(usize),
+    #[error("Border scaling failed. Possible very large border size: {0}")]
+    BorderScalingFailed(u32),
+}
+
+pub trait QrCode {
+    /// QR Codes are strictly squares, so this returns the side length
+    #[must_use]
+    fn side_length(&self) -> usize;
+
+    /// Returns the pixel at coordinate (x, y) if it exists, otherwise None
+    #[must_use]
+    fn pixel(&self, x: usize, y: usize) -> Option<bool>;
+
+    /// If the pixel is out of bounds, return false
+    #[must_use]
+    fn pixel_or_false(&self, x: usize, y: usize) -> bool {
+        self.pixel(x, y).unwrap_or(false)
+    }
+
+    /// Returns the QR code as a vector of booleans, where true represents a filled pixel and false
+    /// represents an empty pixel.
+    /// This contains all the information required to reconstruct the QR code. The side-length
+    /// can be calculated by taking the square root of the length of the vector
+    #[must_use]
+    fn as_vec(&self) -> Vec<bool> {
+        let mut result = Vec::with_capacity(self.side_length() * self.side_length());
+        for y in 0..self.side_length() {
+            for x in 0..self.side_length() {
+                result.push(self.pixel_or_false(x, y));
+            }
+        }
+        result
+    }
+
+    /// Returns a string representation of the QR code, using the given characters as placeholders
+    /// for empty, filled pixels, and the new line character
+    /// border_size is the number of pixels to add around the QR code
+    #[must_use]
+    fn encode_to_string(
+        &self,
+        border_size: u8,
+        empty_char: char,
+        filled_char: char,
+        new_line: char,
+    ) -> String {
+        let mut result = String::with_capacity(2 * self.side_length() * self.side_length());
+        let border: i32 = border_size as i32;
+        for y in -border..self.side_length() as i32 + border {
+            for x in -border..self.side_length() as i32 + border {
+                let c: char = if self.pixel_or_false(x as usize, y as usize) {
+                    filled_char
+                } else {
+                    empty_char
+                };
+                // We push twice because in a standard terminal, chars are rectangular,
```

Comment:
Shouldn't default implemetation return char per pixel and all these console specifics be in a separate ConsoleQrCode or something like that?

Reply:
Maybe... the general case though is available in Vec representation and others. There's no generic string format other than this. This also assumes mono font and things like that. We could rename it to something related to console... that's the best we can do.

Reply:
Maybe it can be a utility function to keep this trait minimalistic. It's just surprising to see console pixel details in a default impl for a trait.

Reply:
I'll rename the function.